util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+
+        img_config['seed'] = check_int(img_config['seed'])
+        log.info('Seed: {0:x}'.format(img_config['seed']))
+        log.info('')
+
+        # Re-initialize with seed to make results reproducible.
+        random.seed(OTP_IMG_SEED_DIVERSIFIER + int(img_config['seed']))
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Adding values to {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+        log.info('')
+        log.info('Parsing OTP image successfully completed.')
+
+    def merge_part_data(self, part):
+        '''This validates and merges the partition data into the memory map dict'''
+
+        part.setdefault('lock', 'false')
+        part['lock'] = check_bool(part['lock'])
+        part.setdefault('items', [])
+
+        if part.setdefault('name', 'unknown_name') == 'LIFE_CYCLE':
+            part.setdefault('state', 'RAW')
+            part.setdefault('count', 'RAW')
+            part['count'] = str(part['count'])
+            if len(part['items']) > 0:
+                raise RuntimeError(
+                    'Life cycle items cannot directly be overridden')
+            if part['lock']:
+                raise RuntimeError('Life cycle partition cannot be locked')
+
+            # Augment life cycle partition with correct life cycle encoding
+            state = self.lc_state.encode('lc_state', part['state'])
+            count = self.lc_state.encode('lc_cnt', part['count'])
+            part['items'] = [{
+                'name': 'LC_STATE',
+                'value': '0x{:X}'.format(state)
+            }, {
+                'name': 'LC_TRANSITION_CNT',
+                'value': '0x{:X}'.format(count)
+            }]
+        else:
+            if len(part['items']) == 0:
+                log.warning("Partition does not contain any items.")
+
+        # Check if partition name exists in memory map and get index
+        part_idx = self.get_part_idx(part['name'])
+        if part_idx < 0:
+            raise RuntimeError('Partition {} does not exist'.format(
+                part['name']))
+
+        # Only partitions with a hardware digest can be locked.
+        if part['lock'] and not \
+           self.config['partitions'][part_idx]['hw_digest']:
+            raise RuntimeError(
+                'Partition {} does not contain a hardware digest'.format(
+                    part['name']))
+
+        # Augment memory map datastructure with lock bit.
+        self.config['partitions'][part_idx]['lock'] = part['lock']
+
+    def merge_item_data(self, part, item):
+        '''This validates and merges the item data into the memory map dict'''
+        item.setdefault('name', 'unknown_name')
+        item.setdefault('value', '0x0')
```

Comment:
Repeating myself, I know, but it's probably worth checking there aren't any stray keys here. It would be a pity to spend hours trying to debug a config that set `valeww` :-)